### PR TITLE
Use proxy function to get PointLocator from MooseMesh

### LINKED ISSUE #7516: Add virtual proxy function for MeshBase::sub_point_locator() to MooseMesh
### Description of the enhancement or error report

I'd like to be able to hook my own point locators into Moose and it seems to me that one way would be by adding a virtual proxy function to `MooseMesh` that by default just returns [`MeshBase::sub_point_locator()`](https://github.com/libMesh/libmesh/blob/a68a631826a813aceeb34ef6b9f44c9be1669d28/src/mesh/mesh_base.C#L500)

### Rationale for the enhancement or information for reproducing the error

I'm using mesh class that constrains the properties of the mesh (orthogonal regular) and I would like to automatically return a custom PointLocator for this mesh (which could exploit guaranteed mesh properties to be _much_ faster). Right now libmesh hardcodes an _all elements tree point locator_. A virtual function could be an elegant way to customize the PointLocator for a Mesh class.

### Identified impact
Will accelerate my LDRD code substantiallty.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #7516

### CODE REVIEW

File: framework/include/vectorpostprocessors/LineMaterialSamplerBase.h [hunk 1]
```diff
@@ -138,19 +138,16 @@ LineMaterialSamplerBase<T>::execute()
   std::vector<Elem *> intersected_elems;
   std::vector<LineSegment> segments;
 
-  MooseSharedPointer<PointLocatorBase> plb = MooseSharedPointer<PointLocatorBase>(_fe_problem.mesh().getMesh().sub_point_locator().release());
-
-  Moose::elementsIntersectedByLine(_start, _end, _fe_problem.mesh(), plb, intersected_elems, segments);
+  UniquePtr<PointLocatorBase> pl = _mesh.getPointLocator();
+  Moose::elementsIntersectedByLine(_start, _end, _mesh, pl, intersected_elems, segments);
```
Comment (dschwen) [inline framework/include/vectorpostprocessors/LineMaterialSamplerBase.h]: @friedmud this looked a bit fishy to me (by maybe you can clarify). You obtained the intersected elements from the undisplaced mesh, but below use the element pointers from that mesh to initialize the (potentially) _displaced_ mesh. I've changed this to consistently use the `_subproblem.mesh()`. Let me know if that is correct.

File: framework/include/postprocessors/PointValue.h [hunk 2]
```diff
@@ -68,7 +68,7 @@ class PointValue : public GeneralPostprocessor
   const VariableValue & _u;
 
   /// A convenience reference to the libMesh::MeshBase object
```
Comment (permcody) [inline framework/include/postprocessors/PointValue.h]: Comment is out of date.

**Comment by permcody** [comment]
I have a general comment: I'd rather stick to the C++11 standard for new methods and parameter declarations in the framework. libMesh returns UniquePtr<> because it is C++03 compliant. We don't need to carry those aliases around in MOOSE though.  We can use `std::unique_ptr<>` directly.

File: framework/src/vectorpostprocessors/ElementsAlongLine.C [hunk 3]
```diff
@@ -47,9 +47,8 @@ ElementsAlongLine::execute()
   std::vector<Elem *> intersected_elems;
   std::vector<LineSegment> segments;
 
-  MooseSharedPointer<PointLocatorBase> plb = MooseSharedPointer<PointLocatorBase>(_fe_problem.mesh().getMesh().sub_point_locator().release());
-
-  Moose::elementsIntersectedByLine(_start, _end, _fe_problem.mesh(), plb, intersected_elems, segments);
+  UniquePtr<PointLocatorBase> pl = _fe_problem.mesh().getPointLocator();
+  Moose::elementsIntersectedByLine(_start, _end, _fe_problem.mesh(), pl, intersected_elems, segments);
```
Comment (permcody) [inline framework/src/vectorpostprocessors/ElementsAlongLine.C]: If we use `std::unique_ptr<>` above does it for us to use std::move here?

Comment (permcody) [inline framework/src/vectorpostprocessors/ElementsAlongLine.C]: NVM   `std::unique_ptr<> &`.... ugh

Comment (dschwen) [inline framework/src/vectorpostprocessors/ElementsAlongLine.C]: `elementsIntersectedByLine` never takes ownership of the pointer, IMO the reference is the right thing for passing this argument. Didn't we have that discussion a few weeks back?

Comment (permcody) [inline framework/src/vectorpostprocessors/ElementsAlongLine.C]: Which is why I advocated for raw pointer in that case. I know I'm not going to win that argument with you 💩  👖 s though so I'll just complain instead.

Comment (permcody) [inline framework/src/vectorpostprocessors/ElementsAlongLine.C]: Luckily respected coders like Herb Sutter, and others have my back:
https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/
https://www.chromium.org/developers/smart-pointer-guidelines
http://timoch.com/blog/2013/04/std-unique_ptr-semantic/
https://books.google.com/books?id=V1kPDAAAQBAJ&pg=PA118&dq=guntheroth+unique_ptr&hl=en&sa=X&ved=0ahUKEwiw1-zBmbXOAhVW92MKHQr6D54Q6AEIHjAA#v=onepage&q=guntheroth%20unique_ptr&f=false

**Comment by dschwen** [comment]
Ok, Herb's got me convinced. The gist being "the callee should not care about the callers object lifetime management strategy". I actually opted to pass as `const PointLocator &` because 

1. it feels more moosey
2. we do not allow null pointers to be passed in (nor do we check for them) 
3. on the caller side we have `*pl` instead of `pl.get()`. I think this looks more "native" and clear.

File: framework/include/utils/RayTracing.h [hunk 4]
```diff
@@ -44,7 +44,7 @@ namespace Moose
  * @param intersected_elems The elements intersected by the line.  Will be empty if there are no intersections.
  * @param segments The line segments across each element
  */
-void elementsIntersectedByLine(const Point & p0, const Point & p1, const MeshBase & mesh, MooseSharedPointer<PointLocatorBase> & point_locator, std::vector<Elem *> & intersected_elems, std::vector<LineSegment> & segments);
+void elementsIntersectedByLine(const Point & p0, const Point & p1, const MeshBase & mesh, const PointLocatorBase & point_locator, std::vector<Elem *> & intersected_elems, std::vector<LineSegment> & segments);
```
Comment (permcody) [inline framework/include/utils/RayTracing.h]: 👍

**Comment by permcody** [comment]
👍

**Pull request merged**